github/table_github_repository_content.go

```diff
@@ -0,0 +1,193 @@
+			{Name: "content", Description: "The decoded file content (if the element is a file).", Type: proto.ColumnType_STRING},
+			{Name: "is_generated", Description: "Whether or not this tree entry is generated.", Type: proto.ColumnType_BOOL},
+			{Name: "is_binary", Description: "Indicates whether the Blob is binary or text.", Type: proto.ColumnType_BOOL},
+			{Name: "sha", Description: "The sha of the file.", Type: proto.ColumnType_STRING, Transform: transform.FromField("CommitUrl").Transform(lastPathElement)},
+			{Name: "commit_url", Description: "Git URL (with SHA) of the file.", Type: proto.ColumnType_STRING},
+		},
+	}
+}
+
+type ContentInfo struct {
+	Oid            string
+	AbbreviatedOid string
+	Name           string
+	Mode           int
+	PathRaw        string
+	IsGenerated    bool
+	Path           string
+	Size           int
+	LineCount      int
+	Type           string
+	Content        string
+	CommitUrl      string
+	IsBinary       bool
+}
+
+//// LIST FUNCTION
+
+func tableGitHubRepositoryContentList(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	owner, repo := parseRepoFullName(d.EqualsQualString("repository_full_name"))
+	var filterPath string
+	if d.EqualsQualString("repository_content_path") != "" {
+		filterPath = d.EqualsQualString("repository_content_path")
+	}
+	plugin.Logger(ctx).Trace("tableGitHubRepositoryContentList", "owner", owner, "repo", repo, "path", filterPath)
+
+	var query struct {
+		RateLimit  models.RateLimit
+		Repository struct {
+			Object struct {
+				Tree struct {
+					Oid            githubv4.String
+					AbbreviatedOid githubv4.String
+					Entries        []struct {
+						Name        githubv4.String
+						Path        githubv4.String
+						Size        githubv4.Int
+						LineCount   githubv4.Int
+						Mode        githubv4.Int
+						PathRaw     githubv4.String
+						IsGenerated githubv4.Boolean
+						Type        githubv4.String
+						Object      struct {
+							Blob struct {
+								Oid            githubv4.String
+								AbbreviatedOid githubv4.String
+								Text           githubv4.String
+								IsBinary       githubv4.Boolean
+								CommitUrl      githubv4.String
+							} `graphql:"... on Blob"`
+							SubTree struct {
+								Entries []struct {
+									Name        githubv4.String
+									Path        githubv4.String
+									Size        githubv4.Int
+									LineCount   githubv4.Int
+									Mode        githubv4.Int
+									PathRaw     githubv4.String
+									IsGenerated githubv4.Boolean
+									Type        githubv4.String
+									Object      struct {
+										Blob struct {
+											Oid            githubv4.String
+											AbbreviatedOid githubv4.String
+											Text           githubv4.String
+											IsBinary       githubv4.Boolean
+											CommitUrl      githubv4.String
+										} `graphql:"... on Blob"`
+									}
+								}
+							} `graphql:"... on Tree"`
```

Comment:
This limits us to one level of directories from the path entered (or repo root), do we want only one level of directories or do we need to figure out how to parse deeper? @cbruno10, thoughts?

Reply:
@cbruno10 / @graza-io, I've delved into a deeper examination of parsing GitHub file contents deeper level. Kindly take a moment to review my findings.

- I attempted to fetch all file contents from a repository by recursively executing a GraphQL query. However, I consistently encountered a `Error: non-200 OK status code: 502 Bad Gateway body` error.
- The analysis was carried out on the `turbot/steampipe-plugin-aws` repository, which contains a significant number of files.
- Despite configuring a rate limiter at the plugin level, I achieved no success.
- I refined the GraphQL query to retrieve file content down to the 3rd, 4th, and 5th levels. Yet, in all scenarios, I faced the same error.
```
Error: non-200 OK status code: 502 Bad Gateway body: "{\n   \"data\": null,\n   \"errors\":[\n      {\n         \"message\":\"Something went wrong while executing your query. This may be the result of a timeout, or it could be a GitHub bug. Please include `04E3:1AE1DB:1814A98:18ECE96:660C236B` when reporting this issue.\"\n      }\n   ]\n}\n" (SQLSTATE HV000)
```
- It may fail due to insufficient storage if the repository has a larger file content.

Based on my observations, attempting to fetch the contents of all files in a repository up to the `nᵗʰ` level tends to be error-prone. On the other hand, we offer flexibility by allowing users to specify the file path for which they wish to obtain content details. By including the `repository_content_path` value in the where clause, we can target the retrieval of file contents from a specific directory within a repository.

I greatly value your feedback and suggestions.

Thank you!

Reply:
@ParthaI Is this behaviour the same compared to when we used the v3 APIs? Are we able to get nested files with that API/the previous table implemented proposed in https://github.com/turbot/steampipe-plugin-github/pull/207?

Reply:
@cbruno10, the modifications in previous PR https://github.com/turbot/steampipe-plugin-github/pull/207 were limited to retrieving file contents at the **root level** only. In the latest PR, we've advanced our approach to parse file contents down to one level down the root.